feat: add new runningUnderARM64Translation property to detect x64 translated apps running on Windows ARM

sergiou87 opened the pull request:
#### Description of Change

This PR adds a new generic property `runningUnderARM64Translation` to `app` that is supposed to supersede `runningUnderRosettaTranslation`, adding support to the same kind of check on Windows: whether or not the app is an x64 build running emulated on a Windows ARM device using [WOW](https://en.wikipedia.org/wiki/Windows_on_Windows). This while keeping the behavior of `runningUnderRosettaTranslation` for macOS.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Added new `app.runningUnderARM64Translation` property to detect when running under Rosetta on Apple Silicon, or WOW on Windows for ARM.

miniak commented on shell/browser/api/electron_api_app.cc:
```diff
@@ -1442,6 +1442,28 @@ void App::SetUserAgentFallback(const std::string& user_agent) {
 
+#if defined(OS_WIN)
+
+typedef BOOL(WINAPI* LPFN_ISWOW64PROCESS2)(HANDLE, PUSHORT, PUSHORT);
```
use this instead
```
  auto IsWow64Process2 =
      reinterpret_cast<decltype(&::IsWow64Process2)>(
          GetProcAddress(GetModuleHandle(L"kernel32.dll"),
                         "IsWow64Process2"));
```

miniak replied:
similar existing code:
https://github.com/electron/electron/blob/c68c65f383f7b210c2ad93715dd70dbb9dcb4798/shell/browser/api/electron_api_power_monitor_win.cc#L47-L50

miniak commented on shell/browser/api/electron_api_app.cc:
```diff
@@ -1442,6 +1442,28 @@ void App::SetUserAgentFallback(const std::string& user_agent) {
+    }
+  }
+
+  return nativeMachine == 0xaa64;
```
no magic values, use proper constant

sergiou87 commented:
Thank you for your feedback @miniak 😄  I addressed all your suggestions and everything is looking good (except the Linux build, but for unrelated reasons I'd say? 😳 )

miniak commented on shell/browser/api/electron_api_app.cc:
```diff
@@ -1442,6 +1442,26 @@ void App::SetUserAgentFallback(const std::string& user_agent) {
+    }
+  }
+
+  return nativeMachine == IMAGE_FILE_MACHINE_ARM64;
```
Shouldn’t you also check if processMachine is not arm64?

sergiou87 replied:
AFAIK that shouldn't be necessary because if both process and OS are arm64, WOW is not used (there is nothing to translate), and therefore `IsWow64Process2` doesn't succeed.

I tested this code with arm64 builds on arm64 machines, x64 builds on arm64 machines and x64 builds on x64 builds on x64 machines, and I got the expected result (only x64 builds on arm64 machines return `true` for this function).

zcbenz commented on docs/api/app.md:
```diff
@@ -1427,11 +1427,26 @@ This is the user agent that will be used when no user agent is set at the
 in your app's initialization to ensure that your overridden value is used.
 
-### `app.runningUnderRosettaTranslation` _macOS_ _Readonly_
+### `app.runningUnderRosettaTranslation` _macOS_ _Readonly_ _Deprecated_
```
Can you add depreciation warning for this property? You can find a few examples searching `EmitWarning` in the codebase.

zcbenz commented on shell/browser/api/electron_api_app.cc:
```diff
@@ -1442,6 +1442,26 @@ void App::SetUserAgentFallback(const std::string& user_agent) {
+  auto IsWow64Process2 = reinterpret_cast<decltype(&::IsWow64Process2)>(
+      GetProcAddress(GetModuleHandle(L"kernel32.dll"), "IsWow64Process2"));
+
+  if (IsWow64Process2) {
```
Should return `false` when `IsWow64Process2` is nullptr.

sergiou87 replied:
It should already be like that, since if it's `nullptr`, `nativeMachine` will remain `0`, different from `IMAGE_FILE_MACHINE_ARM64`. But I'll add an explicit check. Thank you!

zcbenz commented:
API looks good to me.

zcbenz commented:
We have enough approvals from @electron/wg-api here and this should be ready to merge.
